[py] Remove logging API for non-Chromium browsers

cgoldberg opened the pull request:
### **User description**
### 🔗 Related Issues

Fixes #15588 
Fixes #15589
Fixes #15587
Fixes #15586

### 💥 What does this PR do?

This PR changes the logging API in the Python bindings so it only works on Chromium based browsers (Chrome/Edge), as discussed in: https://github.com/SeleniumHQ/selenium/issues/15588

It removes the related methods from the `selenium.webdriver.remote.webdriver.WebDriver` class and adds them to `selenium.webdriver.chromium/webdriver.WebDriver`.

Relevant tests have been updated so the are expected failures on all non-Chromium browsers.

This will bring the Python bindings in line with other language bindings where this has already been removed from most browsers. 

### 🔄 Types of changes
- Breaking change (fix or feature that would cause existing functionality to change)


___

### **PR Type**
Bug fix, Enhancement

titusfortner commented:
So I looked deeper at Ruby Selenium code, and we have a MixIn that allows users to call this method from Chromium. So maybe we just want to move this to Chromium driver class and not delete it entirely (sorry for my earlier mixup).

cgoldberg commented:
OK, I'll update this PR in a little while (and also fix the linting error)

cgoldberg committed:
[py] Moved logging API to Chromium browsers only

cgoldberg committed:
[py] Remove unused import and fix linting errors

cgoldberg commented:
It should be all set now... I updated the PR description also.

I moved the logging methods to the Chromium WebDriver class, so it works in Chrome/Edge. I also adjusted the tests to be expected failures on all browsers except Chrome/Edge.

... just waiting for CI to run.

cgoldberg committed:
Merge branch 'SeleniumHQ:trunk' into py-remove-logging-api

cia2154 commented:
Am I correct in understanding that this PR broke programs that call get_log from webdriver.Remote?
As GraemeWatt mentions, is the only solution to just call driver.execute(Command.GET_LOG, {"type": "<log type>"})["value"]?

The following code works fine in 4.31.0, but doesn't work in 4.32.0.

```
from selenium import webdriver
from selenium.webdriver.remote.command import Command

options = webdriver.ChromeOptions()
options.set_capability('goog:loggingPrefs', {'performance': 'ALL'})

driver = webdriver.Remote(
    command_executor='http://localhost:4444/wd/hub',
    options = options
)

try:
    driver.get('http://example.com')
    # AttributeError: 'WebDriver' object has no attribute 'get_log' in 4.32.0
    print(driver.get_log('performance'))

    # This code works with both versions
    #print(driver.execute(Command.GET_LOG, {"type": "performance"})["value"])

except Exception as e:
    raise e
finally:
    driver.quit()
```

cgoldberg commented:
@cia2154 yes, that is correct. I will look into restoring it for remote chromedriver/edgedriver.

titusfortner commented:
@cgoldberg add those methods to here: https://github.com/SeleniumHQ/selenium/blob/9b48eb13829de163ae66c7dc51890bb8b050031c/py/selenium/webdriver/chromium/remote_connection.py#L45

cgoldberg commented:
@titusfortner I tried to implement this, but I can't find a good way to add `get_log` and `get_log_types` methods to the base `WebDriver` class (for remote) so only Chrome/Edge browsers can access them.

I don't see how adding those methods to `_remote_commands` helps with that.

I think we would need a pretty big restructuring of the class hierarchy to allow Chromium specific methods on a `webdriver.Remote` instance.

Maybe I'm missing something?

cgoldberg commented:
Thinking about this some more...  The only solution I can think of is to move the `get_log` and `get_log_types` methods back to the parent `WebDriver` class, and also add them to subclasses for non-Chromium browsers... and when they are called from a subclass, just raise `NotImplementedError`.

The methods on the base `WebDriver` class would need to check which browser it's using when  those methods are called and raise `NotImplementedError` if the driver isn't Chrome or Edge.

It's kind of ugly, but this would allow the logging API to be used for Chrome, Edge, and Remote drivers, and give a reasonable error when called from a local or remote driver that is not Chrome or Edge.

titusfortner commented:
You don't need to throw a not implemented error. There are a bunch of chromium specific features like casting that work with Remote driver.

This isn't a great way to do it to reference subclasses this way, but it avoids mixins which Python people don't like:
https://github.com/SeleniumHQ/selenium/blob/trunk/py/selenium/webdriver/remote/webdriver.py#L122 

Literally just add those methods to the list above, and webdriver will add the commands to remote sessions that have that browser name

cgoldberg commented:
> There are a bunch of chromium specific features like casting that work with Remote driver.

None of the casting methods (like `get_sink`, `start_tab_mirroring`, `stop_casting`) work on Remote webdrivers (because of the same issue I am running into).

> Literally just add those methods to the list above, and webdriver will add the commands to remote sessions that have that browser name

It doesn't. `_remote_commands` is just a mapping of command names to endpoints. For eaxmple, it contains this:

```
"launchApp": ("POST", "/session/$sessionId/chromium/launch_app")
```

There is a method on `chromium.WebDriver` called `launch_app` that references this command. However, you can't call `launch_app` on a Remote instance... it doesn't exist.

So to use any of the commands listed in `_remote_commands`, you have to implement a method that calls the command. Where do I add that method? Do you have an example of a method you can call on a Remote instance that only works for Chromium browsers? I can't find any.

AlimTleuliyev commented:
Hello! Was this solved? @cgoldberg I am using Remote Driver and I cannot access logs. 

driver.get_log() not avaliable to selenium.webdriver.remote.webdriver

Which version is okay?

cgoldberg commented:
@AlimTleuliyev it was not. Use the workarounds described in https://github.com/SeleniumHQ/selenium/issues/15772
